tls/extensions/s2n_npn.c

```diff
@@ -66,7 +66,38 @@ int s2n_server_npn_recv(struct s2n_connection *conn, struct s2n_stuffer *extensi
         return S2N_SUCCESS;
     }
 
-    POSIX_GUARD_RESULT(s2n_select_server_preference_protocol(conn, extension, supported_protocols));
+    /*
+     *= https://datatracker.ietf.org/doc/id/draft-agl-tls-nextprotoneg-04#section-3
+     *# The "extension_data" field of a "next_protocol_negotiation" extension
+     *# in a "ServerHello" contains an optional list of protocols advertised
+     *# by the server.
+     */
+    if (s2n_stuffer_data_available(extension)) {
+        POSIX_GUARD_RESULT(s2n_select_server_preference_protocol(conn, extension, supported_protocols));
+    }
+
+    /*
+     *= https://datatracker.ietf.org/doc/id/draft-agl-tls-nextprotoneg-04#section-4
+     *# In the event that the client doesn't support any of server's protocols, or
+     *# the server doesn't advertise any, it SHOULD select the first protocol
+     *# that it supports.
+     */
+    if (s2n_get_application_protocol(conn) == NULL) {
+        struct s2n_stuffer stuffer = { 0 };
+        POSIX_GUARD(s2n_stuffer_init(&stuffer, supported_protocols));
+        POSIX_GUARD(s2n_stuffer_skip_write(&stuffer, supported_protocols->size));
+
+        uint8_t length = 0;
+        POSIX_GUARD(s2n_stuffer_read_uint8(&stuffer, &length));
+        POSIX_ENSURE_GT(length, 0);
+
+        uint8_t *data = s2n_stuffer_raw_read(&stuffer, length);
+        POSIX_ENSURE_REF(data);
+
+        POSIX_ENSURE_LT((uint16_t)length, sizeof(conn->application_protocol));
+        POSIX_CHECKED_MEMCPY(conn->application_protocol, data, length);
+        conn->application_protocol[length] = '\0';
```

Comment:
Could you use [s2n_protocol_preferences_read](https://github.com/aws/s2n-tls/blob/main/tls/s2n_protocol_preferences.c#L20-L34) here instead of repeating the parsing logic?

I'm sorry I didn't suggest this before. Just noticed that method :(